Pioneer DDJ-SB mapping

jardiacaj opened the pull request:
I continued my work on the Pioneer DDJ-SB mapping. You can read the forum topic on this, including in page 3 a draft for the corresponding wiki page, here:

http://mixxx.org/forums/viewtopic.php?f=7&t=6886&sid=bab2d5720d8eb823a973cfca92cf5506

This mapping supports all the controller's features, including features blocked in Serato DJ Intro (which is bundled with the controller) and 4 deck control. I found no issues in the latest version of the mapping.

jardiacaj committed:
Pioneer DDJ-SB mapping initial commit

jardiacaj committed:
Pioneer DDJ-SB mapping code quality improvements and beatlooproll leds functionality

jardiacaj committed:
Pioneer DDJ-SB mapping scratching

jardiacaj committed:
Pioneer DDJ-SB mapping improved FX controls

jardiacaj committed:
Pioneer DDJ-SB mapping FX controls fix

jardiacaj committed:
Pioneer DDJ-SB mapping improved LED handling

jardiacaj committed:
Pioneer DDJ-SB mapping added credits to previous authors

jardiacaj committed:
Pioneer DDJ-SB mapping added 4 deck support

jardiacaj committed:
Pioneer DDJ-SB mapping use 14-bit where available

jardiacaj committed:
Pioneer DDJ-SB mapping bugfix deck 2 gain knob

jardiacaj committed:
Pioneer DDJ-SB mapping hot cue pad bugfix

jardiacaj committed:
Pioneer DDJ-SB mapping added LED bindings to FX ch3/4

jardiacaj committed:
Pioneer DDJ-SB mapping improved sync behaviour

jardiacaj committed:
Pioneer DDJ-SB mapping improved binding behaviour

jardiacaj committed:
Pioneer DDJ-SB mapping added some more user options and button functionality

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+// If true the sync button blinks with the beat, if false led is lit when sync is enabled.
+PioneerDDJSB.blinkingSync = false;
+
+// It true, the vinyl button activates slip. Vinyl mode is then activated by using shift.
```
Here's a typo. Change "It" to "if".

jardiacaj replied:
Fixed.

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+//                      CONTROL BINDING                      //
+///////////////////////////////////////////////////////////////
+
+PioneerDDJSB.bulkBindControlConnections = function (group, controlsToFunctions, remove) {
```
Good idea to make this its own function. I'll make use of that in my mapping.

jardiacaj replied:
I moved it for now.

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+        PioneerDDJSB.deckSwitchTable[group] = '[Channel' + deckNumber + ']';
+        PioneerDDJSB.initDeck(PioneerDDJSB.deckSwitchTable[group], false);
+        PioneerDDJSB.setDeckSoftTakeover(PioneerDDJSB.deckSwitchTable[group], true);
+        PioneerDDJSB.setDeckSoftTakeover(PioneerDDJSB.deckSwitchTable[group]);
```
I'm confused about this. Why disconnect soft takeover then enable it again right away for the same deck? Is there really a need to disable soft takeover for any controls? Even if a different deck is active on the controller, I don't see the harm in leaving soft takeover enabled.

jardiacaj replied:
It is the only way I got soft takeover to work. If this is not done like this, soft takeover does not work properly. The first time I switch decks it works, but when switching back the control values change abruptly as if soft takeover was disabled. I don't know if this is the desired behavior. For me it seems like a bug in Mixxx.

The effects need even a worse looking workaround, probably because of the same issue (but turning soft takeover off and on again did not make the trick there for some reason).

Be-ing replied:
If you figure out what the issue is more precisely, please open a bug report on Launchpad.

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+///////////////////////////////////////////////////////////////
+//             TWO MESSAGE DECK INPUT HANDLERS               //
+///////////////////////////////////////////////////////////////
+
```
It would be great if you added a section to the MIDI Scripting wiki page about how to use high resolution MIDI.

jardiacaj replied:
I can do that, it will have to be after my vacation :)

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+    PioneerDDJSB.triggerVinylLed(PioneerDDJSB.channelGroups[group]);
+};
+
+PioneerDDJSB.toggleEngineValue = function (group, control) {
```
Could you move this to common-controller-scripts.js and use the "script" namespace so all scripts can take advantage of it?

sblaisot replied:
IMHO implementing this in common-controller-scripts.js is a good thing but needs to be in a separate PR.

Be-ing commented on res/controllers/Pioneer-DDJ-SB-scripts.js:
```diff
@@ -0,0 +1,888 @@
+///////////////////////////////////////////////////////////////
+
+// Work out the jog-wheel change / delta
+PioneerDDJSB.getJogWheelDelta = function (value) {
```
This is so simple that I don't think it really needs to be a separate function.

Be-ing commented:
Maybe it would be more appropriate to make shift+filter the gain knob rather than shift+effects knob. Gain knobs are usually in mixer sections right by EQ knobs.

Be-ing commented:
It would be nice if you document how this mapping differs from the Serato DJ Intro and Serato DJ mappings.

jardiacaj committed:
Pioneer DDJ-SB mapping minor refactoring

jardiacaj commented:
I also think that making shift+filter the gain knob would be a good idea.

I will be on vacation for the next two weeks and probably won't have an Internet connection at all. I'll gladly continue working on this afterwards, just don't be surprised if for this couple of weeks I'm absent.

Be-ing commented:
I'd prefer to not merge this until the gain control is moved to shift+filter. Other than that and the long function name I commented on, it looks good.

bikedude880 commented:
You can't do shift + filter according to the midi specs. Shift effect is the only logical way to integrate it

Be-ing commented:
Huh? Why wouldn't that work?

bikedude880 commented:
I could be mistaken, but in my research for a 1.12 mapping, the sb midi spec v1.0 doesn't list any shift values for the effects knobs

Be-ing commented:
Oh I didn't realize the controller firmware sends different MIDI values when shift is pressed. Does the shift button send note on/off signals? If so, the script could keep track of whether it is pressed and check that variable in the filter knob handling function.

See http://mixxx.org/wiki/doku.php/midi_scripting#modifier_shift_buttons

bikedude880 commented:
It sends midi note 9n ( n=0-4)  3f.  This shift button modifies values on the board internally as well (performance pads are the primary use)

bikedude880 commented:
Sorry, n=0 or 1 based on the deck

bikedude880 commented:
Been playing around with this mapping, noticed an issue where setting hot cues on deck 3 or 4 will set both cute points for both decks

Be-ing commented:
What exactly do you mean? Setting a hotcue on deck 3 will also set one on deck 1, or it will also set one on deck 4? The hotCueButtons function uses the same method for choosing the deck as every other function handling a deck-specific control, so I don't know why that's happening...

bikedude880 commented:
I'll verify the exact details today after work and ensure I don't have anything else interfering.  It should be deck 3+4 getting duplicated hotcue events

bikedude880 commented:
Hot cue works exactly like it should, failed to notice I loaded the same track in multiple decks.

jardiacaj committed:
Pioneer DDJ-SB mapping gain control with shift+filter knob

jardiacaj committed:
Pioneer DDJ-SB mapping channel fader start

jardiacaj committed:
Renamed script.bindConnections

jardiacaj commented:
Now shift+filter controls pregain. Also the "channel fader start" functionality has been implemented.

Shift+fx knob also allows to control pregain. I can currently not think of a better use for this. Any ideas?

Otherwise I'm quite satisfied with the mapping for now.

Be-ing commented:
Thank you for documenting this on the wiki. The "channel fader start" functionality sounds fun. :)

> Shift+fx knob also allows to control pregain. I can currently not think of a better use for this. Any ideas?

I think mapping it to the effect chain's superknob would make sense.

The method enabling effects on decks 3 + 4 seems quite clunky to me. How about changing the effect enable buttons to apply the effect unit to decks 3 + 4 rather than 1 + 2 if that side of the controller is on deck 3 or 4?

sblaisot commented:
Please don't forget to embed your new controller preset files in windows installer
https://github.com/mixxxdj/mixxx/blob/1.12/build/nsis/Mixxx.nsi#L343

bikedude880 commented:
Love "channel fader start" functionality.  I agree on the effect enable/superknob ideas.

jardiacaj committed:
Pioneer DDJ-SB mapping uninstallation on Windows

jardiacaj committed:
Pioneer DDJ-SB mapping shift + FX knob controls FX super1

jardiacaj committed:
Pioneer DDJ-SB mapping improved FX toggling controls

jardiacaj commented:
I added the files to the Windows installer (but did not test it).

I also made shift + FX knob control super1 instead of gain.

I also changed how effects are enabled/disabled. Now the center button (number 2) toggles the effect for the headphones, and the left and right buttons (1 and 3) toggle the effect for the left and right decks respectively, such that they always control the effects for the active decks (1 or 3 and 2 or 4).

Be-ing commented:
Nice. I think this is ready to merge. :)

Be-ing commented:
Please update the description on the wiki to reflect these changes.

sblaisot commented:
LGTM

Be-ing commented:
Now that you have gone through the process of making this mapping, I'd appreciate your input on the thread on the mailing list about improving the mapping format: http://sourceforge.net/p/mixxx/mailman/mixxx-devel/thread/5588554B.40904%40gmx.com/#msg34230237

bikedude880 commented:
Only left shift works to trigger things like gain, works well on both sides for super knobs.

Deck effects enable is tricky to master since the buttons change their settings when changing decks. Once it is understood how it works it's okay.

jardiacaj commented:
I would like to fix the gain bug. How should this be done? Would it be a new PR?

Be-ing commented:
Yes, now that this PR has been merged, if you want to change something, open a new PR.
